### PR TITLE
feat(internal): Add VECTOR_HOSTNAME env variable

### PR DESCRIPTION
**Pull request by nabokihms**
## Summary
By default, Vector uses the hostname returned by the gethostname. However, in cases when Vector is running in containers, it is meaningless because hostname is always the name of the container, e.g., the name of a pod in Kubernetes cluster.

This commit adds the way to override a hostname by using the env variable. In Kubernetes it can be used like this to see the name of a node where the Vector process is running (which is more meaningful).

```
env:
- name: VECTOR_HOSTNAME
  valueFrom:
    fieldRef:
      fieldPath: spec.nodeName
```

## Change Type
- [ ] Bug fix
- [x] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?
- [ ] Yes
- [x] No

## How did you test this PR?
Tested manually in the kubernetes cluster.

## Does this PR include user facing changes?

- [x] Yes. Please add a changelog fragment based on our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [ ] No. A maintainer will apply the "no-changelog" label to this PR.

## Checklist
- [ ] Please read our [Vector contributor resources](https://github.com/vectordotdev/vector/tree/master/docs#getting-started).
- [ ] If this PR introduces changes Vector dependencies (modifies `Cargo.lock`), please
  run `dd-rust-license-tool write` to regenerate the [license inventory](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv) and commit the changes (if any). More details [here](https://crates.io/crates/dd-rust-license-tool).

## References

Closes https://github.com/vectordotdev/vector/issues/20481

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -198,7 +198,12 @@ pub mod built_info {
 
 /// Returns the host name of the current system.
 pub fn get_hostname() -> std::io::Result<String> {
-    Ok(hostname::get()?.to_string_lossy().into())
+    let hn = option_env!("VECTOR_HOSTNAME").unwrap_or("");
```
Comment (pront) [inline src/lib.rs:201]: Nit: Avoid abbreviations.

```suggestion
    let hostname = option_env!("VECTOR_HOSTNAME").unwrap_or("");
```

**Review by pront — COMMENTED** [review]
Hi @nabokihms, thank you for this enhancement. Please also document this new env var in `website/cue/reference/cli.cue`.

(Tangential to this PR: https://github.com/vectordotdev/vector/issues/21794)

File: changelog.d/21789_add_vector_hostname_env_variable.enhancement.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Add VECTOR_HOSTNAME env variable to override the hostname used in the Vector events and internal metrics.
+This is useful when Vector is running on system where the hostname is not meaningful, e.g., in a container (Kubernetes).
```
Comment (urseberry) [inline changelog.d/21789_add_vector_hostname_env_variable.enhancement.md:2]:
```suggestion
This is useful when Vector is running on a system where the hostname is not meaningful, e.g., in a container (Kubernetes).
```

**Review by urseberry — APPROVED** [review]
Left a suggestion about a grammar nit.

File: src/lib.rs [hunk 3]
```diff
@@ -198,7 +198,12 @@ pub mod built_info {
 
 /// Returns the host name of the current system.
 pub fn get_hostname() -> std::io::Result<String> {
-    Ok(hostname::get()?.to_string_lossy().into())
+    let hostname = option_env!("VECTOR_HOSTNAME").unwrap_or("");
+    Ok(if !hostname.is_empty() {
+        hostname.to_string()
+    } else {
+        hostname::get()?.to_string_lossy().into()
+    })
```
Comment (jszwedko) [inline src/lib.rs:206]: I'd suggest using `VECTOR_HOSTNAME` if set, even if empty, and only fallback to `hostname::get()` if `VECTOR_HOSTNAME` is unset.

**Review by jszwedko — COMMENTED** [review]
Looks good to me! I'll let @pront approve too since he was looking at it before.

I think we could (potentially) add this as a global option in the config file too (`hostname: foo`) but only if the hostname isn't fetched before global configuration is loaded.

**Review by pront — APPROVED** [review]
LGTM. Good call on not treating empty strings values as special.

**Comment by pront** [comment]
> I think we could (potentially) add this as a global option in the config file too (hostname: foo) but only if the hostname isn't fetched before global configuration is loaded.

Would be a nice follow-up to this.

**Review by jszwedko — APPROVED** [review]

**Pull request merged**